wled00/espnow_broadcast.cpp

```diff
@@ -0,0 +1,249 @@
+#define WLED_ESPNOW_WIFI_CHANNEL 1
+
+typedef struct {
+    uint8_t mac[6];
+    uint8_t len;
+    uint8_t data[WLED_ESPNOW_MAX_MESSAGE_LENGTH];
+} QueuedNetworkMessage;
+static_assert(WLED_ESPNOW_MAX_MESSAGE_LENGTH <= ESP_NOW_MAX_DATA_LEN, "WLED_ESPNOW_MAX_MESSAGE_LENGTH must be <= 250 bytes");
+
+ESPNOWBroadcast espnowBroadcast;
+
+ESPNOWBroadcast::ESPNOWBroadcast() {    
+    _networkRxMessages = xRingbufferCreateNoSplit(sizeof(QueuedNetworkMessage), WLED_ESPNOW_MAX_QUEUED_MESSAGES);
+}
+
+bool ESPNOWBroadcast::setup() {
+
+    auto err = esp_event_loop_create_default();
+    if ( ERR_OK != err && ESP_ERR_INVALID_STATE != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_STA_START, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_STA_STOP, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+    err = esp_event_handler_instance_register(WIFI_EVENT, WIFI_EVENT_AP_START, onWiFiEvent, NULL, NULL);
+    if ( ERR_OK != err ) {
+        return false;
+    }
+
+    return setupWiFi();
+}
+
+bool ESPNOWBroadcast::setupWiFi() {
+    Serial.println("ESPNOWBroadcast::setupWiFi()");
+
+    _state.exchange(STOPPED);
+
+    // To enable ESPNow, we need to be in WIFI_STA mode
+    if ( !WiFi.mode(WIFI_STA) ) {
+        Serial.println("WiFi.mode() failed");
+        return false;
+    }
+    // and not have the WiFi connect
+    // Calling discount with tigger an async Wifi Event
+    if ( !WiFi.disconnect(false, true) ) {
+        Serial.println("WiFi.disconnect() failed");
+        return false;
+    }
+
+    return true;
+}
+
+
+void ESPNOWBroadcast::loop(size_t maxMessagesToProcess /*= 1*/) {
+    switch (espnowBroadcast._state.load()) {
+        case ESPNOWBroadcast::STARTING:
+            // if WiFI is in starting state, actually stat ESPNow from our main task thread.
+            start();
+            break;
+        case ESPNOWBroadcast::STARTED: {
+            auto ndx = maxMessagesToProcess;
+            while(ndx-- > 0) {
+                size_t size = 0;
+                auto *msg = (QueuedNetworkMessage*)xRingbufferReceive(_networkRxMessages, &size, 0);
+                if (msg) {
+                    if(size == sizeof(*msg)) {
+                        auto callback = _rxCallback;
+                        while( *callback ) {
+                            (*callback)(msg->mac, msg->data, msg->len);
+                            callback++;
+                        }
+                    }
+                    vRingbufferReturnItem(_networkRxMessages, (void *)msg);
+                } else {
+                    break;
+                }
+            }
+            break;
+        }
+        default:
+            break;
+    }
+}
+
+ESPNOWBroadcast::err_t ESPNOWBroadcast::send(const uint8_t* msg, size_t len) {

```

Comment:
if we do rename the class, then this should be `::broadcast(` as `send` might be unicast in the future

Reply:
this class - by its name is Broadcast. so send implies broadcast. Unicast would be a different implementation / class

Reply:
this is not a huge idealogical point for me, but if you think about it, unicast and broadcast only apply to the sender
the receiver doesn't care which it was. If it gets a message it should listen to, it does.

This class has receiver code in it, therefore it's already something more than broadcast, and less than "everything you would do with ESPNow". If we were to add unicast functionality I wouldn't imagine it being done in `EspNowUnicast` class, with its own receipt hooks.  That's why I'm suggesting this can just be a class called `EspNow`

At the same time, I do understand the single-responsibility principle here and I get what you're going for

Reply:
I could see create a new class which does both broadcast and unicast, and I would implement as you state. This class is an optimized class just for broadcasts as it doesn’t maintain the list of peers and has a small memory footprint